internal/notifier/bitbucketserver.go

```diff
@@ -0,0 +1,281 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package notifier
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	eventv1 "github.com/fluxcd/pkg/apis/event/v1beta1"
+	"github.com/fluxcd/pkg/apis/meta"
+	giturls "github.com/whilp/git-urls"
+)
+
+// BitbucketServer is a notifier for BitBucket Server and Data Center.
+type BitbucketServer struct {
+	ProjectKey      string
+	RepositorySlug  string
+	ProviderUID     string
+	ProviderAddress string
+	Host            string
+	Username        string
+	Password        string
+	Token           string
+	Client          *http.Client
+}
+
+const (
+	buildEndPoint             = "/rest/api/latest/projects/{projectKey}/repos/{repositorySlug}/commits/{commitId}/builds"
+	getBuildStatusQueryString = "key"
+	rqstTimeoutInSeconds      = 5
+)
+
+type RestBuildStatus struct {
+	Name        string `json:"name,omitempty"`
+	Key         string `json:"key,omitempty"`
+	Parent      string `json:"parent,omitempty"`
+	State       string `json:"state,omitempty"`
+	Ref         string `json:"ref,omitempty"`
+	BuildNumber string `json:"buildNumber,omitempty"`
+	Description string `json:"description,omitempty"`
+	Duration    int64  `json:"duration,omitempty"`
+	UpdatedDate int64  `json:"updatedDate,omitempty"`
+	CreatedDate int64  `json:"createdDate,omitempty"`
+	Url         string `json:"url,omitempty"`
+}
+
+type RestBuildStatusSetRequest struct {
+	BuildNumber string `json:"buildNumber,omitempty"`
+	Description string `json:"description,omitempty"`
+	Duration    int64  `json:"duration,omitempty"`
+	Key         string `json:"key"`
+	LastUpdated int64  `json:"lastUpdated,omitempty"`
+	Name        string `json:"name,omitempty"`
+	Parent      string `json:"parent,omitempty"`
+	Ref         string `json:"ref,omitempty"`
+	State       string `json:"state"`
+	Url         string `json:"url"`
+}
+
+// NewBitbucketServer creates and returns a new NewBitbucketServer notifier.
+func NewBitbucketServer(providerUID string, addr string, token string, certPool *x509.CertPool, username string, password string) (*BitbucketServer, error) {
+	hst, id, err := parseBitbucketServerGitAddress(addr)
+	if err != nil {
+		return nil, err
+	}
+
+	comp := strings.Split(id, "/")
+	if len(comp) != 2 {
+		return nil, fmt.Errorf("invalid repository id %q", id)
+	}
+	projectkey := comp[0]
+	reposlug := comp[1]
+
+	httpClient := http.DefaultClient
+	httpClient.Timeout = rqstTimeoutInSeconds * time.Second
+	if certPool != nil {
+		httpClient.Transport = &http.Transport{
+			TLSClientConfig: &tls.Config{
+				RootCAs: certPool,
+			},
+		}
+	}
+
```

Comment:
Yes I did see this client. As much as I wanted to use it, the reason unfortunately I had to drop the idea was that this client only implements a [POST operation](https://github.com/fluxcd/notification-controller/blob/main/internal/notifier/client.go#L84). There is a need for checking existing commit status to avoid unnecessary duplicate commit status updates which is a [GET operation](https://github.com/gdasson/notification-controller/blob/main/internal/notifier/bitbucketserver.go#L179). Now, there are a few ways we can deal with this scenario:

1. Update the existing client to support GET operation and then use this client for both GET and POST operation
2. Make GET call to check duplicate commit status with creating a different client using net/http but use the existing function for POST
3. Ignore this function and make GET and POST call by creating a different client using net/http 
4. Don't make GET call to check duplicate commit status and just post duplicate commit statuses using this function. 

Option 4 is a bad idea of course, so we should ignore that. I went ahead with option 3 currently. Option 2 is feasible but ugly as per my understanding. I think option 1 is the best, and I don't mind submitting additional code in client.go to support GET operation, but that falls outside the scope of this PR.

Please let me know your preference, or any additional option. Happy to implement as suggested. Thx.